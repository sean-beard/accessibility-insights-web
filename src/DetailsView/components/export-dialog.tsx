// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FlaggedComponent } from 'common/components/flagged-component';
import { FeatureFlags } from 'common/feature-flags';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { Dialog, DialogFooter, DialogType, PrimaryButton, TextField } from 'office-ui-fabric-react';
import * as React from 'react';
import { ReportExportServiceProvider } from 'report-export/report-export-service-provider';
import { ExportFormat } from 'report-export/types/report-export-service';
import { ExportResultType } from '../../common/extension-telemetry-events';
import { FileURLProvider } from '../../common/file-url-provider';
import { NamedFC } from '../../common/react/named-fc';
import { DetailsViewActionMessageCreator } from '../actions/details-view-action-message-creator';

export interface ExportDialogProps {
    deps: ExportDialogDeps;
    isOpen: boolean;
    fileName: string;
    description: string;
    html: string;
    onClose: () => void;
    onDescriptionChange: (value: string) => void;
    exportResultsType: ExportResultType;
    onExportClick: () => void;
    featureFlagStoreData: FeatureFlagStoreData;
}

export interface ExportDialogDeps {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
    fileURLProvider: FileURLProvider;
    reportExportServiceProvider: ReportExportServiceProvider;
}

export const ExportDialog = NamedFC<ExportDialogProps>('ExportDialog', props => {
    const [format, setFormat] = React.useState<ExportFormat | null>(null);

    const onDismiss = (): void => {
        props.onClose();
    };

    const onExportLinkClick = (
        event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>,
        exportFormat: ExportFormat,
    ): void => {
        const { detailsViewActionMessageCreator } = props.deps;
        props.onDescriptionChange(props.description);
        detailsViewActionMessageCreator.exportResultsClicked(
            props.exportResultsType,
            props.html,
            event,
        );
        setFormat(exportFormat);
        props.onExportClick();
        props.onClose();
    };

    const onDescriptionChange = (event, value: string): void => {
        props.onDescriptionChange(value);
    };

    const fileURL = props.deps.fileURLProvider.provideURL([props.html], 'text/html');
    const exportService = props.deps.reportExportServiceProvider.forKey(format);
    const ExportForm = exportService ? exportService.exportForm : null;

    const getSingleExportToHtmlButton = () => {
        return (
            <PrimaryButton
                onClick={event =>
                    onExportLinkClick(event as React.MouseEvent<HTMLAnchorElement>, 'download')
                }
                download={props.fileName}
                href={fileURL}
            >
                Export
            </PrimaryButton>
        );
    };

    const getMultiOptionExportButton = () => {
        return (
            <>
                <PrimaryButton
                    text="Export"
                    split
                    splitButtonAriaLabel="Export HTML to any of these format options"
                    aria-roledescription="split button"
                    menuProps={{
                        items: props.deps.reportExportServiceProvider.all().map(service => ({
                            key: service.key,
                            text: service.displayName,
                            onClick: (e: React.MouseEvent<HTMLButtonElement>) => {
                                onExportLinkClick(e, service.key);
                            },
                        })),
                    }}
                    onClick={(e: React.MouseEvent<HTMLAnchorElement>) => {
                        onExportLinkClick(e, 'download');
                    }}
                    download={props.fileName}
                    href={fileURL}
                />
                {ExportForm && (
                    <ExportForm
                        fileName={props.fileName}
                        description={props.description}
                        html={props.html}
                        onSubmit={() => {
                            setFormat(null);
                        }}
                    />
                )}
            </>
        );
    };

    return (
        <Dialog
            hidden={!props.isOpen}
            onDismiss={onDismiss}
            dialogContentProps={{
                type: DialogType.normal,
                title: 'Provide result description',
                subText: 'Optional: please describe the result (it will be saved in the report).',
            }}
            modalProps={{
                isBlocking: false,
                containerClassName: 'insights-dialog-main-override',
            }}
        >
            <TextField
                multiline
                autoFocus
                rows={8}
                resizable={false}
                onChange={onDescriptionChange}
                value={props.description}
                ariaLabel="Provide result description"
            />
            <DialogFooter>
                <FlaggedComponent
                    featureFlag={FeatureFlags.exportReport}
                    featureFlagStoreData={props.featureFlagStoreData}
                    enableJSXElement={getMultiOptionExportButton()}
                    disableJSXElement={getSingleExportToHtmlButton()}
                />
            </DialogFooter>
        </Dialog>
    );
});
