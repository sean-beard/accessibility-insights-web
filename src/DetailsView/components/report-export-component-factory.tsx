// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from 'common/types/visualization-type';
import { CommandBarProps } from 'DetailsView/components/details-view-command-bar';
import {
    ReportExportComponent,
    ReportExportComponentProps,
} from 'DetailsView/components/report-export-component';
import * as React from 'react';

export function getReportExportComponentForAssessment(props: CommandBarProps): JSX.Element {
    const {
        deps,
        assessmentStoreData,
        assessmentsProvider,
        featureFlagStoreData,
        tabStoreData,
    } = props;
    const reportGenerator = deps.reportGenerator;
    const reportExportComponentProps: ReportExportComponentProps = {
        deps: deps,
        exportResultsType: 'Assessment',
        reportGenerator: reportGenerator,
        pageTitle: tabStoreData.title,
        scanDate: deps.getCurrentDate(),
        htmlGenerator: description =>
            reportGenerator.generateAssessmentReport(
                assessmentStoreData,
                assessmentsProvider,
                featureFlagStoreData,
                tabStoreData,
                description,
            ),
        updatePersistedDescription: value =>
            props.deps.detailsViewActionMessageCreator.addResultDescription(value),
        getExportDescription: () => props.assessmentStoreData.resultDescription,
        featureFlagStoreData: props.featureFlagStoreData,
    };

    return <ReportExportComponent {...reportExportComponentProps} />;
}

export function getReportExportComponentForFastPass(props: CommandBarProps): JSX.Element {
    const scanResult = props.visualizationScanResultData.issues.scanResult;

    if (!scanResult) {
        return null;
    }

    const selectedTest = props.visualizationStoreData.selectedFastPassDetailsView;

    if (selectedTest !== VisualizationType.Issues) {
        return null;
    }

    const { deps, tabStoreData } = props;
    const scanDate = deps.getDateFromTimestamp(scanResult.timestamp);
    const reportGenerator = deps.reportGenerator;

    const reportExportComponentProps: ReportExportComponentProps = {
        deps: deps,
        scanDate: scanDate,
        reportGenerator: reportGenerator,
        pageTitle: tabStoreData.title,
        exportResultsType: 'AutomatedChecks',
        htmlGenerator: description =>
            reportGenerator.generateFastPassAutomatedChecksReport(
                scanDate,
                tabStoreData.title,
                tabStoreData.url,
                props.cardsViewData,
                description,
            ),
        updatePersistedDescription: () => null,
        getExportDescription: () => '',
        featureFlagStoreData: props.featureFlagStoreData,
    };

    return <ReportExportComponent {...reportExportComponentProps} />;
}
