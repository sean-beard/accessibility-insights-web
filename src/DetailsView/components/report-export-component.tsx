// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import { escape } from 'lodash';
import { ActionButton } from 'office-ui-fabric-react';
import * as React from 'react';
import { ReportGenerator } from 'reports/report-generator';
import { ExportResultType } from '../../common/extension-telemetry-events';
import { ExportDialog, ExportDialogDeps } from './export-dialog';

export type ReportExportComponentDeps = {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
} & ExportDialogDeps;

export interface ReportExportComponentProps {
    deps: ReportExportComponentDeps;
    exportResultsType: ExportResultType;
    reportGenerator: ReportGenerator;
    pageTitle: string;
    scanDate: Date;
    htmlGenerator: (descriptionPlaceholder: string) => string;
    updatePersistedDescription: (value: string) => void;
    getExportDescription: () => string;
    featureFlagStoreData: FeatureFlagStoreData;
}

export interface ReportExportComponentState {
    isOpen: boolean;
    exportName: string;
    exportDescription: string;
    exportData: string;
}

export class ReportExportComponent extends React.Component<
    ReportExportComponentProps,
    ReportExportComponentState
> {
    constructor(props) {
        super(props);
        this.state = {
            isOpen: false,
            exportName: '',
            exportDescription: '',
            exportData: '',
        };
    }

    private onDismissExportDialog = () => {
        this.setState({ isOpen: false });
    };

    private onExportDescriptionChange = (value: string) => {
        this.props.updatePersistedDescription(value);
        const escapedExportDescription = escape(value);
        this.setState({ exportDescription: escapedExportDescription });
    };

    private generateHtml = () => {
        const { htmlGenerator } = this.props;
        const exportData = htmlGenerator(this.state.exportDescription);
        this.setState({ exportDescription: '', exportData });
    };

    private onExportButtonClick = () => {
        const { reportGenerator, exportResultsType, scanDate, pageTitle } = this.props;
        const exportName = reportGenerator.generateName(exportResultsType, scanDate, pageTitle);
        const exportDescription = this.props.getExportDescription();
        this.setState({ exportDescription, exportName, isOpen: true });
    };

    public render(): JSX.Element {
        const { deps, exportResultsType } = this.props;
        const { isOpen, exportName, exportDescription, exportData } = this.state;
        return (
            <>
                <ActionButton iconProps={{ iconName: 'Export' }} onClick={this.onExportButtonClick}>
                    Export result
                </ActionButton>
                <ExportDialog
                    deps={deps}
                    isOpen={isOpen}
                    fileName={exportName}
                    description={exportDescription}
                    html={exportData}
                    onClose={this.onDismissExportDialog}
                    onDescriptionChange={this.onExportDescriptionChange}
                    exportResultsType={exportResultsType}
                    onExportClick={this.generateHtml}
                    featureFlagStoreData={this.props.featureFlagStoreData}
                />
            </>
        );
    }
}
