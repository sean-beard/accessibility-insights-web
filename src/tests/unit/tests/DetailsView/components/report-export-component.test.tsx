// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import { ActionButton } from 'office-ui-fabric-react';
import * as React from 'react';
import { ReportGenerator } from 'reports/report-generator';
import { IMock, It, Mock, MockBehavior, Times } from 'typemoq';
import { ExportDialog, ExportDialogDeps } from '../../../../../DetailsView/components/export-dialog';
import { ReportExportComponent, ReportExportComponentProps } from '../../../../../DetailsView/components/report-export-component';

describe('ReportExportComponentTest', () => {
    let deps: ExportDialogDeps;
    let props: ReportExportComponentProps;
    let reportGeneratorMock: IMock<ReportGenerator>;
    let htmlGeneratorMock: IMock<(description: string) => string>;
    let updateDescriptionMock: IMock<(value: string) => void>;
    let getDescriptionMock: IMock<() => string>;

    beforeEach(() => {
        deps = {} as ExportDialogDeps;
        reportGeneratorMock = Mock.ofType<ReportGenerator>(undefined, MockBehavior.Strict);
        htmlGeneratorMock = Mock.ofInstance(description => null);
        updateDescriptionMock = Mock.ofInstance(value => null);
        getDescriptionMock = Mock.ofInstance(() => '');
        props = {
            deps,
            exportResultsType: 'Assessment',
            reportGenerator: reportGeneratorMock.object,
            pageTitle: 'test title',
            scanDate: new Date(2019, 5, 28),
            htmlGenerator: htmlGeneratorMock.object,
            updatePersistedDescription: updateDescriptionMock.object,
            getExportDescription: getDescriptionMock.object,
            featureFlagStoreData: {
                'test-feature-flag': true,
            },
        };
    });

    test('render', () => {
        const wrapper = shallow(<ReportExportComponent {...props} />);
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    describe('user interactions', () => {
        test('click export button', () => {
            const persistedDescription = 'persisted description';

            reportGeneratorMock
                .setup(rgm => rgm.generateName(props.exportResultsType, props.scanDate, props.pageTitle))
                .verifiable(Times.once());
            getDescriptionMock
                .setup(gdm => gdm())
                .returns(() => persistedDescription)
                .verifiable(Times.once());
            updateDescriptionMock.setup(udm => udm(It.isValue(persistedDescription))).verifiable(Times.once());
            htmlGeneratorMock.setup(hgm => hgm(It.isAnyString())).verifiable(Times.never());

            const wrapper = shallow(<ReportExportComponent {...props} />);
            const exportButton = wrapper.find(ActionButton);

            exportButton.simulate('click');
            const dialog = wrapper.find(ExportDialog);
            dialog.props().onDescriptionChange(persistedDescription);

            expect(wrapper.getElement()).toMatchSnapshot('dialog should show');

            dialog.props().onClose();

            updateDescriptionMock.verifyAll();
            getDescriptionMock.verifyAll();
            reportGeneratorMock.verifyAll();
            htmlGeneratorMock.verifyAll();
        });

        test('dismiss dialog', () => {
            const wrapper = shallow(<ReportExportComponent {...props} />);
            reportGeneratorMock
                .setup(rgm => rgm.generateName(props.exportResultsType, props.scanDate, props.pageTitle))
                .verifiable(Times.once());
            getDescriptionMock
                .setup(gdm => gdm())
                .returns(() => '')
                .verifiable(Times.once());

            htmlGeneratorMock.setup(hgm => hgm(It.isAnyString())).verifiable(Times.never());

            const exportButton = wrapper.find(ActionButton);
            exportButton.simulate('click');
            const dialog = wrapper.find(ExportDialog);
            dialog.props().onClose();

            expect(wrapper.getElement()).toMatchSnapshot('dialog should be dismissed');
            getDescriptionMock.verifyAll();
            reportGeneratorMock.verifyAll();
            htmlGeneratorMock.verifyAll();
        });

        test('edit text field', () => {
            updateDescriptionMock
                .setup(udm => udm(It.isValue('new description')))
                .returns(() => null)
                .verifiable(Times.once());

            const wrapper = shallow(<ReportExportComponent {...props} />);

            const dialog = wrapper.find(ExportDialog);
            dialog.props().onDescriptionChange('new description');

            expect(wrapper.getElement()).toMatchSnapshot('user input new description');

            updateDescriptionMock.verifyAll();
        });

        test('clicking export on the dialog should trigger the generateHtml', () => {
            const wrapper = shallow(<ReportExportComponent {...props} />);

            htmlGeneratorMock
                .setup(hgm => hgm(wrapper.state('exportDescription')))
                .returns(() => 'test html')
                .verifiable(Times.once());

            const dialog = wrapper.find(ExportDialog);

            dialog.props().onExportClick();

            htmlGeneratorMock.verifyAll();
        });
    });
});
