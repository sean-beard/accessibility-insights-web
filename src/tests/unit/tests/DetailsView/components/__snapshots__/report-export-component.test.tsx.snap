// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportExportComponentTest render 1`] = `
<React.Fragment>
  <CustomizedActionButton
    iconProps={
      Object {
        "iconName": "Export",
      }
    }
    onClick={[Function]}
  >
    Export result
  </CustomizedActionButton>
  <ExportDialog
    deps={Object {}}
    description=""
    exportResultsType="Assessment"
    featureFlagStoreData={
      Object {
        "test-feature-flag": true,
      }
    }
    fileName=""
    html=""
    isOpen={false}
    onClose={[Function]}
    onDescriptionChange={[Function]}
    onExportClick={[Function]}
  />
</React.Fragment>
`;

exports[`ReportExportComponentTest user interactions click export button: dialog should show 1`] = `
<React.Fragment>
  <CustomizedActionButton
    iconProps={
      Object {
        "iconName": "Export",
      }
    }
    onClick={[Function]}
  >
    Export result
  </CustomizedActionButton>
  <ExportDialog
    deps={Object {}}
    description="persisted description"
    exportResultsType="Assessment"
    featureFlagStoreData={
      Object {
        "test-feature-flag": true,
      }
    }
    html=""
    isOpen={true}
    onClose={[Function]}
    onDescriptionChange={[Function]}
    onExportClick={[Function]}
  />
</React.Fragment>
`;

exports[`ReportExportComponentTest user interactions dismiss dialog: dialog should be dismissed 1`] = `
<React.Fragment>
  <CustomizedActionButton
    iconProps={
      Object {
        "iconName": "Export",
      }
    }
    onClick={[Function]}
  >
    Export result
  </CustomizedActionButton>
  <ExportDialog
    deps={Object {}}
    description=""
    exportResultsType="Assessment"
    featureFlagStoreData={
      Object {
        "test-feature-flag": true,
      }
    }
    html=""
    isOpen={false}
    onClose={[Function]}
    onDescriptionChange={[Function]}
    onExportClick={[Function]}
  />
</React.Fragment>
`;

exports[`ReportExportComponentTest user interactions edit text field: user input new description 1`] = `
<React.Fragment>
  <CustomizedActionButton
    iconProps={
      Object {
        "iconName": "Export",
      }
    }
    onClick={[Function]}
  >
    Export result
  </CustomizedActionButton>
  <ExportDialog
    deps={Object {}}
    description="new description"
    exportResultsType="Assessment"
    featureFlagStoreData={
      Object {
        "test-feature-flag": true,
      }
    }
    fileName=""
    html=""
    isOpen={false}
    onClose={[Function]}
    onDescriptionChange={[Function]}
    onExportClick={[Function]}
  />
</React.Fragment>
`;
